src/js/content_scripts/gamepad-navigator.js

```diff
@@ -0,0 +1,119 @@
+
+Licensed under the BSD 3-Clause License. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the BSD 3-Clause License at
+https://github.com/fluid-lab/gamepad-navigator/blob/master/LICENSE
+*/
+
+/* eslint-env browser */
+
+(function (fluid) {
+    "use strict";
+
+    var gamepad = fluid.registerNamespace("gamepad");
+
+    fluid.defaults("gamepad.gamepadNavigator", {
+        gradeNames: ["fluid.modelComponent"],
+        model: {
+            index: null,
+            axes: [],
+            buttons: []
+        },
+        listeners: {
+            "onCreate.gamepadListener": "{that}.gamepadListener"
+        },
+        frequency: 100,
+        connectivityIntervalReference: null,
+        invokers: {
+            gamepadListener: {
+                funcName: "gamepad.gamepadNavigator.gamepadListener",
+                args: ["{that}"]
+            },
+            transformArrayToObject: {
+                funcName: "gamepad.gamepadNavigator.arrayToObject",
+                args: ["{arguments}.0"]
+            }
+        }
+    });
+
+    /**
+     *
+     * Listens to the connected and disconnected events of the gamepad and updates the
+     * gamepad navigator component's model according to the state of the gamepad.
+     *
+     * @param {Object} that - The gamepad navigator component.
+     *
+     */
+    gamepad.gamepadNavigator.gamepadListener = function (that) {
+        // When a gamepad is connected
+        window.addEventListener("gamepadconnected", function () {
+            // Store the gamepad info if no other gamepad is already connected.
+            if (!fluid.isValue(that.model.index)) {
+                // Scan the state of gamepad frequently.
+                that.options.connectivityIntervalReference = setInterval(function () {
+                    // Catch the current state of gamepad.
+                    var connectedGamepad = navigator.getGamepads()[0];
+
+                    // Initiate the gamepad navigator model transaction.
+                    var modelUpdateTransaction = that.applier.initiate();
+
+                    modelUpdateTransaction.fireChangeRequest({ path: "index", value: connectedGamepad.index });
+                    modelUpdateTransaction.fireChangeRequest({ path: "axes", value: that.transformArrayToObject(connectedGamepad.axes) });
+                    modelUpdateTransaction.fireChangeRequest({ path: "buttons", value: that.transformArrayToObject(connectedGamepad.buttons) });
+
+                    // Commit the current state of gamepad.
+                    modelUpdateTransaction.commit();
+                }, that.options.frequency);
+            }
+        });
+
+        // When gamepad is disconnected
+        window.addEventListener("gamepaddisconnected", function (event) {
+            // Ensure that the gamepad navigator doesn't break if any other gamepad is disconnected.
+            if (that.model.index === event.gamepad.index) {
```

Comment:
So, if someone wants to connect a different controller, they have to ensure that the first controller is detected as disconnected first, and then connect the second controller.  Otherwise they've already missed their chance to use the second controller.  It seems to me like you're probably going to want to handle this differently, either handling multiple controllers separately, or handling inputs from all detected controllers.  One thought I had was to create a dynamic component for each controller when it's connected, and then to destroy that when the controller is detected as having disconnected.

Reply:
Hi Tony, I started thinking of the flaw in my approach after I read this comment. As you said, in this approach, the user will miss the chance to use the second controller unless they connect it after disconnecting the first controller. So I tried a different approach.

When the user disconnects the primary controller they're using, the extension will get the list of gamepads and the first connected gamepads it finds in the list will be used to serve as the new primary controller. I think with this approach, we won't need to handle multiple controllers and their inputs at once. At the same time, we are offering the users the flexibility to immediately and automatically switch to another gamepad if their current controller fails.

I was able to do this with a bit of restructuring in the current code. Moreover, I think it'll also work as we restructure the gamepad event listener using the approach you suggested in the previous comment. Please let me know what you think and I'll carry this out as we will finalize.

Reply:
So I tried to set up the gamepad events wrapper along with the approach I mentioned above, and it seems to work as expected. I can update the pull with that if my approach (mentioned in the previous comment) sounds good to you.